src/dtx/dtx_rpc.c

```diff
@@ -705,11 +735,18 @@ dtx_rpc_prep(struct ds_cont_child *cont,d_list_t *dti_list,  struct dtx_entry **
 	}
 
 	/* Use helper ULT to handle DTX RPC if there are enough helper XS. */
-	if (dss_has_enough_helper())
-		rc = dss_ult_create(dtx_rpc_helper, dca, DSS_XS_IOFW, dca->dca_tgtid, 0,
-				    &dca->dca_helper);
-	else
-		rc = dtx_rpc_internal(dca);
+	if (dss_has_enough_helper()) {
+		rc = ABT_eventual_create(0, &dca->dca_chore_eventual);
+		if (rc != ABT_SUCCESS) {
+			D_ERROR("failed to create eventual: %d\n", rc);
+			rc = dss_abterr2der(rc);
+			goto out;
+		}
+		rc = dss_chore_delegate(&dca->dca_chore, dtx_rpc_helper);
+	} else {
+		dss_chore_diy(&dca->dca_chore, dtx_rpc_helper);
```

Comment:
It may be better to hide dss_chore_diy(), instead, just call dss_chore_delegate() and check whether there is enough helper XS inside dss_chore_delegate (your patch has already does similar check, may need enhancement), then determine:
1. Add chore to the queue if possible. Or
2. Start ULT on neighbor to help me. Or
3. Directly do it by myself.
With such approach, then DTX logic needs not to understand helper XS knowledge. And also save one chore API.

Reply:
The problem is that in case 1 and 2, the API call returns asynchronously (i.e., the chore may or may not have started and terminated), whereas in case 3, the API call returns only if the chore has started and terminated. The user code needs to differentiate between these two scenarios. Therefore I feel two API functions for these two scenarios are clearer and thus simpler.

Reply:
The caller still calls dtx_rpc_post() without any changes to distinguish the cases.

Reply:
I prefer not to merge `diy` with `delegate`, if that's acceptable to you.

Reply:
Current API can work. Mine is just for suggestion.